### PR TITLE
Pgcopy stream metrics

### PR DESCRIPTION
**Pull request by oleurud**
In order to measure the behavior of the copy endpoints (following https://github.com/CartoDB/cartodb-platform/issues/4414)

Things that are interesting to be measured: 
- query time, 
- size of payload, 
- format (text, csv, or binary)

### CODE REVIEW

File: app/utils/query_info.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+        copyQuery = copyQuery.toUpperCase();
+
+        if (!copyQuery.toUpperCase().startsWith("COPY ")) {
+            return false;
+        }
+       
+        const regex = /(\bFORMAT\s+)(\w+)/;
+        const result = regex.exec(copyQuery);
+        if (result && result.length >= 3 && COPY_FORMATS.includes(result[2])) {
```
Comment (rafatower) [inline app/utils/query_info.js]: Minor thing: most likely you don't need a `(capture group)` for the `\bFORMAT\s+`. Also, you don't expect more than 1 format in the query string. So, you could write

```js
        const regex = /\bFORMAT\s+(\w+)/;
        const result = regex.exec(copyQuery);
        if (result && result.length === 2 && COPY_FORMATS.includes(result[1])) {
            //...
        }
```

File: app/utils/query_info.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+            format = result[2];
+        }
```
Comment (rafatower) [inline app/utils/query_info.js]: The default format, if not present in options is `text`. I guess we want to have it in the stats for simplicity.

File: test/unit/query_info.test.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            });
+        });
+
+        describe('text', function() {
+            const csvValidQueries = [
+                "COPY copy_endpoints_test (id, name) FROM STDIN WITH (FORMAT TEXT)",
+                "COPY copy_endpoints_test (id, name) FROM STDIN",
+            ];
+
+            csvValidQueries.forEach(query => {
```
Comment (rafatower) [inline test/unit/query_info.test.js]: > csvValidQueries

csv or text ?

File: app/utils/query_info.js (hunk 2 above)
Comment (rafatower) [reply app/utils/query_info.js]: ah, sorry, did not see the default above!

File: app/utils/query_info.js (hunk 1 above)
Comment (oleurud) [reply app/utils/query_info.js]: Nice, thank you!

File: test/unit/query_info.test.js (hunk 3 above)
Comment (oleurud) [reply test/unit/query_info.test.js]: :scream:  copy paste

**Comment by oleurud** [comment]
Reviewers, take into account ---> https://github.com/CartoDB/cartodb-platform/issues/4414#issuecomment-390932252

File: app/controllers/copy_controller.js [hunk 4]
```diff
@@ -55,99 +57,132 @@ CopyController.prototype.route = function (app) {
...
+            size: 0,
+            time: null,
+            format: getFormatFromCopyQuery(req.query.sql)
+        };
+
+        res.header("Content-Disposition", `attachment; filename=${encodeURIComponent(filename)}`);
+        res.header("Content-Type", "application/octet-stream");
+
+        try {
+            const start_time = Date.now();
```
Comment (enekid) [inline app/controllers/copy_controller.js]: `startTime` :)

File: app/controllers/copy_controller.js [hunk 5]
```diff
@@ -55,99 +57,132 @@ CopyController.prototype.route = function (app) {
...
+                    const end_time = Date.now();
+                    res.body = {
+                        time: (end_time - start_time) / 1000,
+                        total_rows: copyFromStream.rowCount
+                    };
+
+                    return next();
+                });
+
+                if (req.get('content-encoding') === 'gzip') {
```
Comment (enekid) [inline app/controllers/copy_controller.js]: hopefully we can handle this gunziping with nginx, but at the momment it's ok

File: app/utils/query_info.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+const COPY_FORMATS = ['TEXT', 'CSV', 'BINARY'];
```
Comment (enekid) [inline app/utils/query_info.js]: If the user specifies an invalid format (eg: `WADUS`), we will set the format to `TEXT`. Shouldn't we return an error, or at least a `false`

File: app/controllers/copy_controller.js (hunk 4 above)
Comment (oleurud) [reply app/controllers/copy_controller.js]: :+1:

File: app/controllers/copy_controller.js (hunk 5 above)
Comment (oleurud) [reply app/controllers/copy_controller.js]: https://github.com/CartoDB/cartodb-platform/issues/4447

File: app/utils/query_info.js (hunk 6 above)
Comment (oleurud) [reply app/utils/query_info.js]: If the user sends a bad format, the request will fail. 

But for sure, this module should work independently. I am going to change it

Comment (oleurud) [reply app/utils/query_info.js]: I have made some changes... not perfect, but enough for me

**Pull request merged**